### ISSUE TITLE
Support labels for board sizes from 25 to 52

### ISSUE THREAD

**Issue by abaisero** [julianandrews/sgf-render#43]
Hi!

First of all thank you for this very useful tool;  I've found myself using it a couple different times in the past few years.

I'm currently working on a small project that generates QR codes and renders them as go boards (in part inspired by the fact that QR codes were themselves inspired by Go), and I'm hoping to use sgf-render in this pipeline for the rendering part.  The only issue I'm having right now is that QR codes are often bigger than what sgf-render supports.

The SGF file format allows (for Go) board sizes up to 52 (taking into account both lowercase and uppercase letters), whereas sgf-render only seems to support board sizes up to 26 (only taking into account lowercase letters).
Though QR codes can get much bigger than that, it would be already a huge boost in usability to be able to use sgf-render with the full range up to size 52.
Would it be possible to extend sgf-render to fulfill the whole SGF range?
Unfortunately I know nothing of rust, or else I'd try to submit a pull request of my own;  but hopefully this should be a simple extension?

Best,

**Comment by abaisero** [comment]
Just a quick follow-up;

I just realized that there might be two issues at hand here.

If I try to generate a sgf with size larger than 26, I get "Invalid move", which I think is due to the problem mentioned earlier, that the program only allows ranges up to 26.

But actually I also noticed that I get a different error for board size of exactly 26, "Range too large for use with labels", due to the fact that the labels skip letter I and therefore only supports up to size 25. Weirdly, this also happens if I use the --no-labels option. I think that would also need to be addressed if full boards up to 52 are to be supported.

**Comment by abaisero** [comment]
Ah, nevermind that last comment about the --no-labels option;  I mixed up --no-labels with --no-board-labels. 
--no-board-labels works as intended.  (though it would be nice to allow board labels to work on the wider ranges as well, e.g., using AA, AB, AC, or just numbers on both sides.  But that's a minor secondary concern.

**Comment by julianandrews** [comment]
It's great to see people finding new uses for this! Thanks for reporting the issue.

I'm glad you found `--no-board-labels` - that's the first thing I was going to suggest. If that works well for you, then great!

The issue with the larger SGF and "Invalid move" is definitely some sort of bug. If the file is valid SGF, it should be possible to render it. I'll open a new issue to investigate that.

I would like to support labels for non-standard board sizes up to 52x52, but since the convention is to skip 'I' we only get to 51 at most, which is awkward. I don't really want to increase the supported range without reaching the SGF spec 52x52. There's also the awkwardness that the labels along the top have to be a single character, while the labels along the sides are often 2 characters, so command line flags for configuring them need to be a little different. Anyway, I think this is a real problem that needs solving, but when I've tried in the past I've failed to come up with a design I was actually happy with.

I'll definitely look into the invalid move issue in the next few days. The labeling issue is something I want to fix, but need a good design for before I do it.

**Comment by julianandrews** [comment]
@abaisero - would you be willing to share the SGF file you used that gave a "Invalid move" error? It would really help to reproduce the issue.

**Comment by abaisero** [comment]
Ah!  ok, new developments.

The good news is that I was still using 3.1.5, and the "invalid move" issue does not arise when I upgrade to 3.2.0. Sorry about that, my bad.

The bad news is that there is still an error, and the bigger goban sizes are still not handled correctly. Judging by the rendering, it seems like the program is still interpreting the upper-case coordinates A, B, C, etc as if they were lower-case, a, b, c, etc.

Here's an example SGF that should fill the entire board with stones, but if you render it you can see that the last 3 rows and 3 columns are actually empty. This should also represent a QR code, and you can see that certain markups that should exist in a QR code (the "eyes") don't actually appear because they are being overwritten due to the above issue.

Here's the SGF:

```
(;GM[1]FF[4]SZ[29]AW[ah][ai][ak][al][am][ap][aq][at][au][av][bb][bc][bd][be][bf][bh][bj][bk][bl][bm][bn][bp][bq][bt][bv][bx][by][bz][bA][bB][cb][cf][ch][ck][cl][cm][cn][cr][ct][cv][cx][cB][db][df][dh][do][dp][dq][dt][du][dv][dx][dB][eb][ef][eh][ei][ek][el][en][ep][ev][ex][eB][fb][fc][fd][fe][ff][fh][fi][fj][fk][fl][fo][fp][fs][ft][fu][fv][fx][fy][fz][fA][fB][gh][gj][gl][gn][gp][gr][gt][gv][ha][hb][hc][hd][he][hf][hg][hh][hk][hl][hm][hn][hr][ht][hu][hv][hw][hx][hy][hz][hA][hB][hC][ib][ic][id][ie][if][ih][ik][il][io][ip][iq][is][iu][ix][iy][iC][jb][jf][jg][jh][jj][jm][jq][ju][jw][jz][jC][ka][kb][kd][ke][kf][kh][kj][kk][kl][kq][kr][kt][ku][kw][kz][kA][kB][kC][lb][lc][ld][lg][lh][li][lk][lm][ln][lr][ls][lt][lx][ly][lA][lB][lC][ma][mc][md][mk][mp][mr][ms][mu][mv][my][mz][mA][mB][nb][ne][nf][ng][ni][nj][nl][no][nv][nz][nA][oa][ob][oc][of][oj][ol][os][ot][ov][ow][oB][oC][pa][pb][pc][pd][pf][pg][pi][pj][pm][pp][pq][pr][ps][pt][px][pz][pB][qe][qh][qi][qm][qn][qp][qq][qu][qv][qw][qx][qy][qB][qC][rb][rd][rf][rg][rh][ri][rk][rm][rn][ro][rq][rz][sc][sd][se][sf][sh][sm][sq][sr][ss][su][sy][sA][sB][tb][td][te][tf][tg][th][tj][tk][tl][tm][tn][tq][tr][ts][tv][tx][tz][tA][tB][tC][ub][uc][ud][uf][uh][ui][uk][ul][um][un][uo][up][uq][us][ut][uz][va][vb][vc][vd][ve][vf][vg][vh][vl][vo][vq][vr][vt][vv][vw][vx][vA][vB][vC][wh][wi][wv][wx][wB][wC][xb][xc][xd][xe][xf][xh][xi][xj][xo][xp][xr][xv][xw][xx][xz][xA][yb][yf][yh][yi][ym][yo][yq][yr][ys][yA][yB][zb][zf][zh][zi][zk][zm][zn][zo][zr][zt][zw][zx][zy][zB][Ab][Af][Ah][Ai][Aj][Al][Ao][Aq][Ar][AC][Bb][Bc][Bd][Be][Bf][Bh][Bi][Bn][Bo][Bp][Bq][Br][Bs][Bt][Bu][Bw][BB][Ch][Ck][Cl][Cn][Cq][Cs][Ct][Cu][Cv][Cx][Cz][CB][CC]AB[aa][ab][ac][ad][ae][af][ag][aj][an][ao][ar][as][aw][ax][ay][az][aA][aB][aC][ba][bg][bi][bo][br][bs][bu][bw][bC][ca][cc][cd][ce][cg][ci][cj][co][cp][cq][cs][cu][cw][cy][cz][cA][cC][da][dc][dd][de][dg][di][dj][dk][dl][dm][dn][dr][ds][dw][dy][dz][dA][dC][ea][ec][ed][ee][eg][ej][em][eo][eq][er][es][et][eu][ew][ey][ez][eA][eC][fa][fg][fm][fn][fq][fr][fw][fC][ga][gb][gc][gd][ge][gf][gg][gi][gk][gm][go][gq][gs][gu][gw][gx][gy][gz][gA][gB][gC][hi][hj][ho][hp][hq][hs][ia][ig][ii][ij][im][in][ir][it][iv][iw][iz][iA][iB][ja][jc][jd][je][ji][jk][jl][jn][jo][jp][jr][js][jt][jv][jx][jy][jA][jB][kc][kg][ki][km][kn][ko][kp][ks][kv][kx][ky][la][le][lf][lj][ll][lo][lp][lq][lu][lv][lw][lz][mb][me][mf][mg][mh][mi][mj][ml][mm][mn][mo][mq][mt][mw][mx][mC][na][nc][nd][nh][nk][nm][nn][np][nq][nr][ns][nt][nu][nw][nx][ny][nB][nC][od][oe][og][oh][oi][ok][om][on][oo][op][oq][or][ou][ox][oy][oz][oA][pe][ph][pk][pl][pn][po][pu][pv][pw][py][pA][pC][qa][qb][qc][qd][qf][qg][qj][qk][ql][qo][qr][qs][qt][qz][qA][ra][rc][re][rj][rl][rp][rr][rs][rt][ru][rv][rw][rx][ry][rA][rB][rC][sa][sb][sg][si][sj][sk][sl][sn][so][sp][st][sv][sw][sx][sz][sC][ta][tc][ti][to][tp][tt][tu][tw][ty][ua][ue][ug][uj][ur][uu][uv][uw][ux][uy][uA][uB][uC][vi][vj][vk][vm][vn][vp][vs][vu][vy][vz][wa][wb][wc][wd][we][wf][wg][wj][wk][wl][wm][wn][wo][wp][wq][wr][ws][wt][wu][ww][wy][wz][wA][xa][xg][xk][xl][xm][xn][xq][xs][xt][xu][xy][xB][xC][ya][yc][yd][ye][yg][yj][yk][yl][yn][yp][yt][yu][yv][yw][yx][yy][yz][yC][za][zc][zd][ze][zg][zj][zl][zp][zq][zs][zu][zv][zz][zA][zC][Aa][Ac][Ad][Ae][Ag][Ak][Am][An][Ap][As][At][Au][Av][Aw][Ax][Ay][Az][AA][AB][Ba][Bg][Bj][Bk][Bl][Bm][Bv][Bx][By][Bz][BA][BC][Ca][Cb][Cc][Cd][Ce][Cf][Cg][Ci][Cj][Cm][Co][Cp][Cr][Cw][Cy][CA])
```

**Comment by abaisero** [comment]
For additional reference, this is what the above SGF renders as (ignore the extra padding;  I added that in post because QR codes need additional border around:

![image](https://github.com/user-attachments/assets/50d99052-35cf-46ec-82da-1065c2efd579)

**Comment by julianandrews** [comment]
Ok! The issues with high coordinates should be fixed in v3.2.1:

```
$ sgf-render --version
sgf-render 3.2.1
$ echo '(;GM[1]FF[4]SZ[30:4];B[bb]W[Bc])' | sgf-render --format text --no-board-labels
┏┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┯┓
┠○┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┨
┠┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼┼●┼┨
┗┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┷┛
```

Give that a try and let me know if you have any other problems. It's really useful to have someone poking at the edge cases with the large boards - that hasn't gotten enough testing.

I'm going to leave this issue open to track the labeling issue, but I won't be fixing that until I've got a clear plan for what the labeling and CLI interface will look like. My current assumption is that basically no one is affected, so it's not super high priority. Do let me know if you actually need the labels though so I can prioritize appropriately!

**Comment by julianandrews** [comment]
Scanning through my old issues, I've realized this is now basically a duplicate of #17 (with the actual bugs now theoretically fixed). So I'm going to close this issue in favor of the original, and add some context there for clarity.
